synapse/storage/databases/state/bg_updates.py

```diff
@@ -98,81 +98,84 @@ def _get_state_groups_from_groups_txn(
-                            WHERE {where_clause}
-                            ORDER BY type, state_key, state_group DESC
-                        )
-                        """
+        # This is an optimization to create a select clause per-condition. This
+        # makes the query planner a lot smarter on what rows should pull out in the
+        # first place and we end up with something that takes 10x less time to get a
+        # result.
+        use_condition_optimization = (
+            not state_filter.include_others and not state_filter.is_full()
+        )
+        state_filter_condition_combos: List[Tuple[str, Optional[str]]] = []
+        # We only need to caclculate this list if we're using the condition optimization
+        if use_condition_optimization:
+            for etype, state_keys in state_filter.types.items():
+                if state_keys is None:
+                    state_filter_condition_combos.append((etype, None))
+                else:
+                    for state_key in state_keys:
+                        state_filter_condition_combos.append((etype, state_key))
+        # And here is the optimization itself. We don't want to do the optimization
+        # if there are too many individual conditions. 10 is an arbitrary number
+        # with no testing behind it but we do know that we specifically made this
+        # optimization for when we grab the necessary state out for
+        # `filter_events_for_client` which just uses 2 conditions
+        # (`EventTypes.RoomHistoryVisibility` and `EventTypes.Member`).
+        if use_condition_optimization and len(state_filter_condition_combos) < 10:
+            select_clause_list: List[str] = []
+            for etype, skey in state_filter_condition_combos:
+                if skey is None:
+                    where_clause = "(type = ?)"
+                    overall_select_query_args.extend([etype])
+                else:
+                    where_clause = "(type = ? AND state_key = ?)"
+                    overall_select_query_args.extend([etype, skey])
+
+                # We could use `SELECT DISTINCT ON` here to align with the query below
+                # but that isn't compatible with SQLite and we can get away with `LIMIT
+                # 1` here instead because the `WHERE` clause will only ever match and
+                # target one event; and is simpler anyway. And it's better to use
+                # something that's simpler and compatible with both Database engines.
+                select_clause_list.append(
+                    f"""
+                    (
+                        SELECT type, state_key, event_id
+                        FROM state_groups_state
+                        INNER JOIN sgs USING (state_group)
+                        WHERE {where_clause}
+                        ORDER BY type, state_key, state_group DESC
+                        LIMIT 1
                     )
```

Comment:
SQLite doesn't like parenthesis around each clause: `(select_clause_A) UNION (select_clause_B)` -> `sqlite3.OperationalError: near "(": syntax error`

And if we simply remove the parenthesis, it only allows the `ORDER BY` and `LIMIT` to be at the end which throws off our `LIMIT 1` trick.

Related links:

 - https://stackoverflow.com/questions/4653124/what-does-the-sql-standard-say-about-parentheses-in-sql-union-except-intersect-s

Reply:
Fixed this with a `wrap_union_if_postgres(...)` function where we only add the `LIMIT 1` in the Postgres case as it's really just an optimization and then `ORDER BY state_group DESC` after the `UNION` which is compatible with SQLite and postgres -> https://github.com/matrix-org/synapse/pull/14531/commits/a790a45671f99c74509da9f85fe620688d467208

Getting conscious that some of these subtle hacks might not be better than a big `if-else` distinction between Postgres and SQLite :grimacing: